src/backend/joanie/payment/models.py

```diff
@@ -1 +1,463 @@
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.core.validators import MaxValueValidator, MinValueValidator
+from django.db import models
+from django.dispatch import receiver
+from django.utils import timezone
+from django.utils.functional import cached_property
+from django.utils.translation import get_language
+from django.utils.translation import gettext_lazy as _
+
+from babel.numbers import get_currency_symbol
+from djmoney.models.fields import MoneyField
+from djmoney.money import Money
+from howard.issuers import InvoiceDocument
+from parler.utils.context import switch_language
+
+from joanie.core.models import Order
+from joanie.core.utils import merge_dict
+
+from . import enums as payment_enums
+from . import get_payment_backend
+
+User = get_user_model()
+
+
+class Invoice(models.Model):
+    """
+    Invoice model is an accounting element related to an order
+    """
+
+    parent = models.ForeignKey(
+        "self",
+        on_delete=models.RESTRICT,
+        related_name="children",
+        verbose_name="parent",
+        blank=True,
+        null=True,
+    )
+    reference = models.CharField(
+        db_index=True,
+        editable=False,
+        max_length=40,
+        unique=True,
+    )
+    created_on = models.DateTimeField(
+        _("created on"),
+        auto_now_add=True,
+        editable=False,
+    )
+    updated_on = models.DateTimeField(
+        _("updated on"),
+        auto_now=True,
+        editable=False,
+    )
+    order = models.OneToOneField(
+        Order,
+        verbose_name=_("order"),
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
+    )
+    amount = MoneyField(
+        _("amount"),
+        max_digits=9,
+        decimal_places=2,
+        default_currency=settings.DEFAULT_CURRENCY,
+        null=False,
+        blank=False,
+    )
+    recipient_name = models.CharField(
+        _("invoice recipient"), max_length=40, null=False, blank=False
+    )
+    recipient_address = models.TextField(
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(models.Q(order__isnull=False) & models.Q(parent__isnull=True))
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def root_parent(self):
```

Comment:
Weird. Shouldn't you forbid a credit not to have children? so invoice.parent.parent should always be None if invoice.parent is not None?

Reply:
From your point of view, an order should have an invoice and invoice could have only credit note as children. Am I right?

In my mind a credit_note could have invoice as children. In this way, if staff user did a mistake when he created credit_note, he will be able to create a related invoice to correct his mistake and we are able to see the relation from admin interface.

Reply:
that wayn, if you made an error on a credit note, you can fix it with a sibling invoice?

Reply:
Yes, sibling invoice is also fine and we enlight logic.